Brooks/LanguageRenderer/NewCSharpRenderer.hs

```diff
@@ -0,0 +1,561 @@
+  equals, semi, vcat, lbrace, rbrace, render, colon, render)
+
+newtype CSharpCode a = CSC {unCSC :: a}
+
+instance Functor CSharpCode where
+    fmap f (CSC x) = CSC (f x)
+
+instance Applicative CSharpCode where
+    pure = CSC
+    (CSC f) <*> (CSC x) = CSC (f x)
+
+instance Monad CSharpCode where
+    return = CSC
+    CSC x >>= f = f x
+
+instance PackageSym CSharpCode where
+    type Package CSharpCode = ([(Doc, Label, Bool)], Label)
+    packMods n ms = liftPairFst (sequence ms, n)
+
+instance RenderSym CSharpCode where
+    type RenderFile CSharpCode = (Doc, Label, Bool)
+    fileDoc code = liftTripFst (liftA3 fileDoc' (top code) (fmap tripFst code) bottom, tripSnd $ unCSC code, tripThird $ unCSC code)
+    top _ = liftA2 cstop endStatement (include "")
+    bottom = return empty
+
+instance KeywordSym CSharpCode where
+    type Keyword CSharpCode = Doc
+    endStatement = return semi
+    endStatementLoop = return empty
+
+    include _ = return $ text "using"
+    inherit = return colon
+
+    list _ = return $ text "List"
+    listObj = return new
+
+    blockStart = return lbrace
+    blockEnd = return rbrace
+
+    ifBodyStart = blockStart
+    elseIf = return $ text "else if"
+    
+    iterForEachLabel = return $ text "foreach"
+    iterInLabel = return $ text "in"
+
+    commentStart = return doubleSlash
+    
+    printFunc = return $ text "Console.Write"
+    printLnFunc = return $ text "Console.WriteLine"
+    printFileFunc = fmap (printFileDocD "Write")
+    printFileLnFunc = fmap (printFileDocD "WriteLine")
+
+instance PermanenceSym CSharpCode where
+    type Permanence CSharpCode = Doc
+    static = return staticDocD
+    dynamic = return dynamicDocD
+
+instance BodySym CSharpCode where
+    type Body CSharpCode = Doc
+    body = liftList bodyDocD
+    bodyStatements = block
+    oneLiner s = bodyStatements [s]
+
+    addComments s = liftA2 (addCommentsDocD s) commentStart
+
+instance BlockSym CSharpCode where
+    type Block CSharpCode = Doc
+    block sts = (lift1List blockDocD endStatement (map (fmap fst) (map state sts)))
+
+instance StateTypeSym CSharpCode where
+    type StateType CSharpCode = Doc
+    bool = return $ boolTypeDocD
+    int = return $ intTypeDocD
+    float = return $ csFloatTypeDoc
+    char = return $ charTypeDocD
+    string = return $ stringTypeDocD
+    infile = return $ csInfileTypeDoc
+    outfile = return $ csOutfileTypeDoc
+    listType p st = liftA2 listTypeDocD st (list p)
+    intListType p = listType p int
+    floatListType p = listType p float
+    boolListType = listType dynamic bool
+    obj t = return $ typeDocD t
+    enumType t = return $ typeDocD t
+    iterator _ = error "Iterator-type variables do not exist in C#"
+
+instance ControlBlockSym CSharpCode where
+    runStrategy l strats rv av = 
+        case Map.lookup l (Map.fromList strats) of Nothing -> error $ "Strategy '" ++ l ++ "': RunStrategy called on non-existent strategy."
```

Comment:
I don't like super-long lines, I would prefer to use a new line and less indent.  Also, a local function using `where` could be good here.

Reply:
Style fixes in 7e619d656288976863ec517a37896b3c6427d26a and 54f17c8450fc6193959a07ceeadf329eb7cf5028. Simplified this function in c1e48fb5503d21482c789a131932270e748dee33.